### PR TITLE
[Framework] Update Musical Chairs to capture edge case

### PR DESCRIPTION
**Pull request by 0xzoz**
The current implementation of the stop_the_music function in the musical_chairs module had a logic flaw that allowed the current_seats to increment more than it should. Specifically, the function was designed to increment the current_seats by 1 under certain conditions related to network compliance, but it did so even when the number of compliant nodes was equal to current_seats.

The changes fix the logic to ensure that current_seats only increments by 1 more than the number of compliant nodes, thereby maintaining the integrity of the validator set and ensuring fair participation.

### CODE REVIEW

**Comment by 0o-de-lally** [comment]
Great turnaround on this! Rockstar. I think there's a decision ahead.

Here's how I think of it:
Every epoch we offer (O ) number of seats.
We may fill all of them (F), but we also may not.
Then the number of seats in the epoch is always F.
So when we decide to increase the number of seats by +1, this should be F+1 not O+1.
The bug happens because if assumes assumes F==O (meaning there's always more demand for seats than is offered).

I think what you are suggesting is similar. Where you are incrementing Performant nodes (P). P will also be about equal to F in the happy cases. Though it can be 5% lower than F. Did I understand that correctly?

If we want musical_chairs::Chairs.current_seats to be P, then what you are doing is perfect. If we want it to be F, what needs to happen is that  when seats (F) are filled in the epoch prologue ( epoch_boundary.process_incoming_validators )  we call out to musical_chairs to set the number (a function for this does not yet exist AFAIK).

**Comment by 0xzoz** [comment]
Thanks for the context @0o-de-lally. As per the conversation in discord, for this to use F, we check the number filled seats after PoF auction and then adjust the `chairs.current_seats` for when `musical_chairs.stop_the_music()` is executed at the next epoch boundary.

File: framework/libra-framework/sources/ol_sources/musical_chairs.move [hunk 1]
```diff
@@ -64,38 +64,43 @@ module ol_framework::musical_chairs {
...
-          // remove chairs
-          // reduce the validator set to the size of the compliant set.
-          chairs.current_seats = vector::length(&compliant);
+        let num_compliant_nodes = vector::length(&compliant);
+        let compliance_ratio = fixed_point32::multiply_u64(100, *&ratio);
+
+        // Conditions under which seats should be one more than the number of compliant nodes(<= 5%)
+        if fixed_point32::is_zero(*&ratio) || (compliance_ratio <= 5) {
+            // Correct the number of seats if it is not one more than the number of compliant nodes
+            if chairs.current_seats != chairs.current_seats + 1 {
```
[the remark this answers is not in the document]
Comment (0o-de-lally) [reply framework/libra-framework/sources/ol_sources/musical_chairs.move]: Please take a look at the `!=`

**Review by 0o-de-lally — APPROVED** [review]
Woot!

**Pull request merged**